### ISSUE TITLE
Remaining TODOs

### ISSUE THREAD

**Issue by philbucher** [KratosMultiphysics/CoSimIO#25]
left to do:
- [x] finish C exposure of the `Info` => #23 
- [x] after #23: update the C-Interface to work again 
- [x]  add enums for connection-status [as @pooyan-dadvand requested](https://github.com/KratosMultiphysics/CoSimIO/pull/14#discussion_r422144343)
- [x] Add tutorials for Python
- [x] Add tutorials for C
- [x] after #13: Make sure the tutorials are up to date
- [x] check what to do with IDs of nodes & elements (start from 0 or from 1?) => 1, same as in Kratos
- [x] Merge of settings and info in cpp and python tutorials
- [x] Making ctests running the python integration tests (see this link) => py-tests are not run through CTest, bcs they should not go through Valgrind
- [x] in py-tutorial: Explain how to use different py-versions: https://pybind11.readthedocs.io/en/stable/compiling.html#configuration-variables => #88
- [x] in C-Interface: Some improvements needed in Im-/ExportMesh => #49
- [ ] improve python-interface, such that it doesn't copy all the time
- [x] ~for python: construct `Info` from `dict`~ => doesn't really work
- [x] enable C-Tests in Win
- [x] passing tests by valgrind
- [ ] Keep track of created Info in C to delete the undeleted one at the end
- [x] implement missing tests for `Info` in `C` => #53
- port missing tests from Kratos:
    - [ ] Communication
    - [x] DataContainer
- [x] implement `begin` & `end` in `DataContainer` to make them usable in range-based loops
- [ ] Document the settings used for each IO method. In the error message when setting is invalid or empty?
- [x] Making the import data size only input and return the imported size in Info? 

#
patch version `string` vs `int`
- according to https://semver.org/ it should be MAJOR(`int`).MINOR(`int`).PATCH(`int`)
    => every small change increments the PATCH
    => users should use the releases and not the development branch
- every time we hardcode the patch version (string) into the code it is a commit. this is a bit messy
- Getting the automatic update of the patches to work is a bit messy
#
Should `Import` & `Export` always take two arguments? =>`Info` and an object containing the data? => e.g. a `Mesh` object that could take e.g. only nodes, but optionally also Elements & IDs => this way we wouldn't need to change the interface if we want to change sth.
I think this would be easier to manage the memory too, i.e. where to allocate/free it

Those objects could take the "native" data-structures or provide their own one such that the memory doesn't need to be copied.
For Python this is kinda necessary anyway

e.g. for `Mesh` this could look like the following:
~~~c++
class Mesh
{
    void SetNodalCoordinates(CoSimIO::DataContainer);
    void SetNodeIds(CoSimIO::DataContainer);
    void SetElements(CoSimIO::DataContainer ElementConnectivities, CoSimIO::DataContainer ElementTypes);
    void SetElementIds(CoSimIO::DataContainer ElementIds);
    void PrintVTK();
};

Same concept could be used for `Geometry` and possibly others in the future

**Comment by pooyan-dadvand** [comment]
I like the idea of mesh object and I think we should go in this direction. Nevertheless, I would not postpone the v1 for this change which is rather large.

**Comment by philbucher** [comment]
@pooyan-dadvand I agree, I already opened #55 for collecting the things that should be done for the next release

**Comment by philbucher** [comment]
closing as all the remaining TODOs are listed in #55
